vm-setup/roles/v1aX_integration_test/tasks/node_reuse.yml

```diff
@@ -182,175 +223,247 @@
+    register: machines
     retries: 200
     delay: 20
-    until: provisioned_machine.stdout|int == 2
+    until:
+      - machines is succeeded
+      - machines.resources | filter_phase("running") | length == 2
+
+  - name: Get provisioned BMH before upgrade in MachineDeployment node reuse test scenario.
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: saved_bmhs
 
   - name: Get provisioned BMH names and UUIDs mapping before upgrade in MachineDeployment node reuse test scenario.
-    shell: |
-        kubectl get bmh -A -o json | jq '.items[]| select (.status.provisioning.state == "provisioned")
-        | "metal3/"+.metadata.name+"="+"metal3://"+.metadata.uid' | cut -f2 -d\" | sort > /tmp/before_upgrade_mapping_md.txt
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    set_fact:
+      bmh_before_upgrade_md: "{{ bmh_before_upgrade_md | default([]) + [item.metadata.name + '=metal3://' + item.metadata.uid] }}"
+    with_items: "{{ saved_bmhs.resources | filter_provisioning('provisioned') }}"
 
   - name: Update maxSurge and maxUnavailable fields in MachineDeployment.
-    shell: |
-        kubectl patch -n "{{ NAMESPACE }}" machinedeployment "{{ CLUSTER_NAME }}" --type=merge -p '{"spec": {"strategy": {"rollingUpdate": {"maxSurge": 0,"maxUnavailable": 1}}}}'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    k8s:
+      api_version: cluster.x-k8s.io/{{ CAPI_VERSION }}
+      kind: MachineDeployment
+      name: "{{ CLUSTER_NAME }}"
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+      definition:
+        spec:
+          strategy:
+            rollingUpdate:
+              maxSurge: 0
+              maxUnavailable: 1
 
   - name: Update Metal3MachineTemplate nodeReuse field to 'True'.
-    shell: |
-        kubectl get m3mt "{{ CLUSTER_NAME }}"-workers -n "{{ NAMESPACE }}" -ojson | jq '.spec.nodeReuse=true' | kubectl apply -f-
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-
-  - name: List only ready BMHs'.
-    shell: kubectl get bmh -n "{{ NAMESPACE }}" -o json | jq -r '[ .items[] | select (.status.provisioning.state == "ready") | .metadata.name ]'
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: ready_bmhs
+    k8s:
+      api_version: infrastructure.cluster.x-k8s.io/{{ CAPM3_VERSION }}
+      kind: Metal3MachineTemplate
+      name: "{{ CLUSTER_NAME }}-workers"
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+      definition:
+        spec:
+          nodeReuse: True
+
+  - name: List BMHs'.
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: bmhs
```

Comment:
you should list only **Ready** BMHs' here, not all of them, are not you? Because at this point you should have 2 of them and those need to be marked unhealthy

Reply:
Because I could not register bmhs after filter this being done in the next task so I am marking only Ready bmhs as unhealthy

Reply:
Okay got it